src/sources.js

```diff
@@ -286,6 +286,62 @@ class CommunityNarrative extends Narrative {
   }
 }
 
+
+class UrlDefinedSource extends Source {
+
+  static get _name() { return "urlDefined"; }
+  get baseUrl() {
+    throw new Error("UrlDefinedSource does not use `this.baseUrl`");
+  }
+
+  dataset(pathParts) {
+    return new UrlDefinedDataset(this, pathParts);
+  }
+  narrative(pathParts) {
+    return new UrlDefinedNarrative(this, pathParts);
+  }
+
+  // what does this mean in the context of this source?
+  async availableDatasets() { return []; }
+
+  async availableNarratives() { return []; }
+
+  async getInfo() { return {}; } // todo
+
+}
+
+class UrlDefinedDataset extends Dataset {
+  get baseParts() {
+    return this.pathParts;
+  }
+  get isRequestValidWithoutDataset() {
+    return false;
+  }
+  baseNameFor(type) {
+    if (type==="main" || type==="tree" || type==="meta") {
```

Comment:
The logic within `parsePrefix` calls the `urlFor` method for "main", "tree" and "meta" datasets before any requests are made. An alternate solution would be to modify `urlFor` to return `undefined` for the latter two types, and therefore `baseNameFor` would only be valid if `type==="main"`

Reply:
94f91a5 improves this function (and the `urlFor` method)